R/class.R

```diff
@@ -56,20 +72,25 @@ PyClass <- function(classname, defs = list(), inherit = NULL) {
   defs <- lapply(defs, function(x) {
     if (inherits(x, "function")) {
       f <- inject_super(x)
-      x <- function(...) {
-        args <- list(...)
-        # enable convertion scope for `self`
-        # the first argument is always `self`.and we don't want to convert it.
-        assign("convert", TRUE, envir = as.environment(args[[1]])) 
-        do.call(f, append(args[1], lapply(args[-1], py_to_r)))
-      }
+      x <- enable_convert_scope(f)
+      attr(x, "__env__") <- environment(f)
     }
     x
   })
   
-  bt$type(
+  type <- bt$type(
     classname, inherit, 
     do.call(reticulate::dict, defs)
   )
+  
+  # we add a reference to the type here. so it can be acessed without needing
+  # to find the type from self.
+  lapply(defs, function(x) {
+    if(!is.null(e <- attr(x, "__env__"))) {
+      e$`__class__` <- type
```

Comment:
Doesn't this cause some potential confusion between the real Python `__class__` attribute, and the thing we're putting in this environment?

Reply:
I used the name `__class__` following what python seems to use. For example if you run:

```
class Base:
  def __init__ ():
    print("Hello Base")
    

MyClass = type(
  "MyClass",
  (Base,),
  {
    "__init__": lambda x: super()
  }
)

a = MyClass()
```

You will get: `RuntimeError: super(): __class__ cell not found`.

Reply:
Got it. I was confused whether this was distinct from Python's own `__class__` object; but it's clear now that the intention is to ensure the Python `__class__` object is set with an appropriate object.